### PR TITLE
Implement digestMultibase.

### PR DESCRIPTION
**Pull request by dmitrizagidulin**
(Test vector obtained from a modified js-multiformats lib)

### CODE REVIEW

File: lib/digest.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @param {object} options - Options hashmap.
+ * @param {object} options.input - JSON-LD object.
+ * @param {string} [options.canonAlgorithm] - Canonicalization algorithm.
+ * @param {Function} options.documentLoader - JSON-LD document loader.
+ *
+ * @returns {Promise<string>} Returns the canonhash digest, base58-encoded.
+ */
+export async function multibaseDigest({
+  input, canonAlgorithm = URDCA_2015_ALG, documentLoader
```
Comment (dlongley) [inline lib/digest.js]:
```suggestion
  input, canonizeAlgorithm = URDCA_2015_ALG, documentLoader
```

File: lib/digest.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+} = {}) {
+  let normalized;
+
+  switch(canonAlgorithm) {
+    case URDCA_2015_ALG:
+      normalized = await jsonld.canonize(input, {documentLoader});
```
Comment (dlongley) [inline lib/digest.js]: Probably needs these two options:

https://github.com/digitalbazaar/jsonld-signatures/blob/02d0f63d7aa75a619233482ce9d0d4098057a5b2/lib/suites/LinkedDataSignature.js#L194-L196

Comment (dmitrizagidulin) [reply lib/digest.js]: @dlongley - I was thinking about that, initially. Are you sure we'd want those? We're specifically defining our minimal `multibaseDigest` as "canonize to N-Quads, using URDNA2015". I'd vote for leaving out the options until we need them.

Comment (dlongley) [reply lib/digest.js]: @dmitrizagidulin, I don't mean allow someone to customize those options, I mean you have to set them. I don't think the default output is n-quads.

File: README.md [hunk 3]
```diff
@@ -47,7 +47,16 @@ npm install
...
 
-TBD
+```js
+import {multibaseDigest} from '@digitalbazaar/minimal-digest';
+// or
+const {multibaseDigest} = require('@digitalbazaar/minimal-digest');
+
+const data = {key: 'value'};
+// defaults to sha-256 hash, 'urdca2015' canonicalization for objects, base58btc encoding
+await multibaseDigest({data, documentLoader});
```
Comment (dlongley) [inline README.md]:
```suggestion
const result = await multibaseDigest({data, documentLoader});
```

File: lib/digest.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+/*!
+ * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
+ */
+import jsonld from 'jsonld';
+import {encode as base58Encode} from 'base58-universal';
+import {sha256digest} from './sha256digest.js';
+
+const URDCA_2015_ALG = 'urdca2015';
+// Multicodec table entry code 0xb403, converted to varint
+const URDCA2015_HEADER = new Uint8Array([131, 232, 2]);
```
Comment (dlongley) [inline lib/digest.js:10]: I'm just presuming this is right. :)

File: lib/digest.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
+/*!
+ * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
+ */
+import jsonld from 'jsonld';
+import {encode as base58Encode} from 'base58-universal';
```
Comment (dlongley) [inline lib/digest.js]: Alphabetize:

```suggestion
import {encode as base58Encode} from 'base58-universal';
import jsonld from 'jsonld';
```

File: lib/digest.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Performs a combined canonicalize + multihash operation.
+ *
+ * @param {object} options - Options hashmap.
+ * @param {object} options.input - JSON-LD object.
+ * @param {string} [options.canonizeAlgorithm] - Canonicalization algorithm.
+ * @param {Function} options.documentLoader - JSON-LD document loader.
+ *
+ * @returns {Promise<string>} Returns the canonhash digest, base58-encoded.
```
Comment (dlongley) [inline lib/digest.js]: `canonhash` seems weird? Is that intentional?

```suggestion
 * @returns {Promise<string>} Returns the canonical hash digest, base58-encoded.
```

File: lib/sha256digest-browser.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * @param {string} string - Input string to hash.
+ *
+ * @returns {Uint8Array} Hash digest bytes.
+ */
+export async function sha256digest({string}) {
+  const bytes = new TextEncoder().encode(string);
+  return new Uint8Array(
+    await crypto.subtle.digest('SHA-256', bytes)
+  );
```
Comment (dlongley) [inline lib/sha256digest-browser.js]:
```suggestion
  return new Uint8Array(await crypto.subtle.digest('SHA-256', bytes));
```

File: lib/sha256digest.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * Hashes a string of data using SHA-256.
+ *
+ * @param {string} string - Input string to hash.
+ *
+ * @returns {Uint8Array} Hash digest bytes.
+ */
+export async function sha256digest({string}) {
+  return new Uint8Array(
+    crypto.createHash('sha256').update(string).digest()
+  );
```
Comment (dlongley) [inline lib/sha256digest.js]:
```suggestion
  return new Uint8Array(crypto.createHash('sha256').update(string).digest());
```

File: lib/sha256digest-browser.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
+ */
+/* eslint-env browser */
+const crypto = self && (self.crypto || self.msCrypto);
```
Comment (dlongley) [inline lib/sha256digest-browser.js]: Should this be `globalThis` not `self` now?

File: test/digest.spec.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
```
Comment (dlongley) [inline test/digest.spec.js]:
```suggestion
 * Copyright (c) 2021-2022 Digital Bazaar, Inc. All rights reserved.
```

File: test/loader.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+// import didContext from 'did-context';
+import ed25519 from 'ed25519-signature-2020-context';
+// import x25519 from 'x25519-key-agreement-2020-context';
+import cred from 'credentials-context';
+import ageContext from 'age-verification-context';
+import {JsonLdDocumentLoader} from 'jsonld-document-loader';
+
+// const didKeyDriver = didKey.driver();
+// const resolver = new CachedResolver();
+// resolver.use(didKeyDriver);
```
Comment (dlongley) [inline test/loader.js]: Can we remove all this commented stuff?

**Review by dlongley — APPROVED** [review]
Approving with several nits, see comments. Thanks!

File: lib/digest.js (hunk 4 above)
Comment (dmitrizagidulin) [reply lib/digest.js:10]: @dlongley - hehe, I got it from https://github.com/multiformats/js-multiformats

File: lib/sha256digest-browser.js (hunk 9 above)
Comment (dmitrizagidulin) [reply lib/sha256digest-browser.js]: Changed.

**Pull request merged**